play: Add prompt choice to importer

diomekes opened the pull request:
This is for #2008.

I hope my refactoring isn't way out of line.

jrobeson commented:
I fixed the appveyor issue. Can you rebase this on masterand push again?

sampsyo commented:
This is a cool idea; thanks! Any chance you could add some documentation on the feature?

The refactoring also looks fine on first glance (I haven't done a full review yet). Anything in particular I should be looking for? (Or perhaps you can just give a brief overview of what needed changing?)

diomekes commented:
Yes, I forgot about adding documentation, I'll do that.

I just defined separate methods/functions so that the CLI and event listener could access them.  I did have to change the way the command is called for the importer prompt since the plugin originally would exit once the command was executed, but I needed it to return to beets.  All my tests had it working fine.

sampsyo commented on beetsplug/play.py:
```diff
@@ -117,35 +111,76 @@ def play_music(self, lib, opts, args):
+        """Returns either the raw paths of items or a playlist of the items.
+        """
+        raw = config['play']['raw'].get(bool)
+        if raw:
```
This can be shortened to:

```py
if config['play']['raw']:
```

sampsyo commented on beetsplug/play.py:
```diff
@@ -117,35 +111,76 @@ def play_music(self, lib, opts, args):
+
+    def _print_info(self, selection, command_str, open_args,
+                    item_type='track'):
+        """Prompts user whether to continue if playlist exceeds threshold.
```
I know this is minor, but docstrings should typically be written in the imperative voice—here, "Prompt" rather than "Prompts."

sampsyo replied:
Also, this docstring should probably explain what the return value means. Here, if I understand correctly, it returns true if the operation should be aborted (i.e., it returns false if the operation should actually happen, which is a little counterintuitive).

sampsyo commented on beetsplug/play.py:
```diff
@@ -117,35 +111,76 @@ def play_music(self, lib, opts, args):
-        else:
-            open_args = [self._create_tmp_playlist(paths)]
-
         self._log.debug(u'executing command: {} {!r}', command_str, open_args)
```
Maybe these logging messages should go in the method that actually plays the music?

sampsyo commented on beetsplug/play.py:
```diff
@@ -117,35 +111,76 @@ def play_music(self, lib, opts, args):
 
             if ui.input_options((u'Continue', u'Abort')) == 'a':
-                return
+                return True
```
The method should explicitly `return False` at the end (when it doesn't return True).

sampsyo commented on beetsplug/play.py:
```diff
@@ -155,3 +190,21 @@ def _create_tmp_playlist(self, paths_list):
         m3u.close()
         return m3u.name
+
+    def before_choose_candidate_listener(self, session, task):
```
Argh; I wish we had chosen a better name than `before_choose_candidate`. :scream: But that's hardly your fault.

sampsyo commented on docs/plugins/play.rst:
```diff
@@ -29,6 +29,18 @@ would on the command-line)::
+adds a `plaY` option to the prompt, so pressing `y` will execute the configured
+command and play the items currently being imported.
+
+Once you exit your configured player, you will be returned to the import
```
Perhaps more accurately, "once the configured command exists, …"

sampsyo commented on beetsplug/play.py:
```diff
@@ -117,35 +111,76 @@ def play_music(self, lib, opts, args):
+        if not cancel:
+            play(command_str, paths, open_args)
+
+    def _create_command_str(self, args=None):
```
This can probably be called just `_command_str`.

sampsyo commented on beetsplug/play.py:
```diff
@@ -19,17 +19,35 @@
+    """
+    try:
+        if keep_open:
+            command = command_str.split()
```
I'd recommend using `beets.util.shlex_split` instead of `str.split` for more predictable command spitting behavior.

diomekes commented:
Thanks for the thorough review, I think I addressed all the issues and hopefully the warning_threshold configuration option handling is a little more sophisticated.

sampsyo commented:
This looks great; thank you! I don't know why the 3.6 test failed the most recent time around, but I'm going to merge this now. To simplify things, I may remove that old, backwards-compatible misspelling…
